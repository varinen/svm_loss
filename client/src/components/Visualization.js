import React, {Component} from 'react';
import MathJax from 'react-mathjax';
import {connect} from 'react-redux';
import {getPlot} from '../actions/visualization'


const formulas = {
    f1: `W_{0,0}x_0+W_{0,1}x_1+b_0`,
    f2: `(x_0,x_1)`,
    f3: `(W_{0,0},W_{0,1})`
};

export class Visualization extends Component {
    render() {
        return (
            <MathJax.Provider>
                <div className="visualization card shadow">
                    <div className="card-header py-3">
                        <h6 className="m-0 font-weight-bold text-primary">
                            Data And Classifier Visualization
                        </h6>
                    </div>
                    <div className="card-body">
                        <article>
                            <p>
                                Data points are shown as circles colored by
                                their
                                class (red/green/blue). The background regions
                                are
                                colored by whichever class is most likely at
                                any
                                point according to the current weights.
                            </p>

                            <p>
                                Each classifier is visualized by a line that
                                indicates its zero score level set. For
                                example,
                                the blue classifier computes scores
                                as <MathJax.Node inline
                                                 formula={formulas.f1}/> and
                                the blue line shows the set of
                                points <MathJax.Node inline
                                                     formula={formulas.f2}/> that
                                give score of zero. The blue arrow draws
                                the
                                vector <MathJax.Node inline
                                                     formula={formulas.f3}/> ,
                                which shows the direction of score increase and
                                its
                                length is proportional to how steep the
                                increase
                                is.
                            </p>
                        </article>

                        <article>
                            <div className="mb-4">
                                <img alt="Plot" id="plot" className="img-fluid"
                                     src={`data:image/png;base64,${this.props.visualization.plot}`}/>
                            </div>
                        </article>

                        <hr/>

                    </div>
                </div>
            </MathJax.Provider>
        );
    }
}

const mapStateToProps = state => {
    const {visualization, data, params} = state;
    return {visualization, data, params};
};

export default connect(mapStateToProps, {getPlot})(Visualization);