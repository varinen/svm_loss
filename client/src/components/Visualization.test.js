import React from 'react';
import {shallow} from 'enzyme';
import {Visualization} from './Visualization';
import configureMockStore from "redux-mock-store";

const mockStore = configureMockStore();
const store = mockStore({});


describe('Visualization', () => {
    const props = {
        plot: 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP09l5cDwAD3AG60PCC3wAAAABJRU5ErkJggg=='
    };

    const vis = shallow(<Visualization {...props}/>);

    it('renders correctly', () => {
        expect(vis).toMatchSnapshot();
        expect(vis.find('img').length).toEqual(1);
        expect(vis.find('#plot').prop('src')).toEqual(props.plot)
    });
});

