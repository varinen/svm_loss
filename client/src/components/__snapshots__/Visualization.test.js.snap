// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visualization renders correctly 1`] = `
<MathJaxProvider
  options={
    Object {
      "showMathMenu": false,
      "showMathMenuMSIE": false,
      "tex2jax": Object {
        "inlineMath": Array [],
      },
    }
  }
  script="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.4/MathJax.js?config=TeX-MML-AM_CHTML"
>
  <div
    className="visualization card shadow"
  >
    <div
      className="card-header py-3"
    >
      <h6
        className="m-0 font-weight-bold text-primary"
      >
        Data And Classifier Visualization
      </h6>
    </div>
    <div
      className="card-body"
    >
      <article>
        <p>
          Data points are shown as circles colored by their class (red/green/blue). The background regions are colored by whichever class is most likely at any point according to the current weights.
        </p>
        <p>
          Each classifier is visualized by a line that indicates its zero score level set. For example, the blue classifier computes scores as 
          <MathJaxNode
            formula="W_{0,0}x_0+W_{0,1}x_1+b_0"
            inline={true}
          />
           and the blue line shows the set of points 
          <MathJaxNode
            formula="(x_0,x_1)"
            inline={true}
          />
           that give score of zero. The blue arrow draws the vector 
          <MathJaxNode
            formula="(W_{0,0},W_{0,1})"
            inline={true}
          />
           , which shows the direction of score increase and its length is proportional to how steep the increase is.
        </p>
      </article>
      <article>
        <div
          className="mb-4"
        >
          <img
            alt="Plot"
            className="img-fluid"
            id="plot"
            src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP09l5cDwAD3AG60PCC3wAAAABJRU5ErkJggg=="
          />
        </div>
      </article>
      <hr />
    </div>
  </div>
</MathJaxProvider>
`;
